state.go

```diff
@@ -529,6 +561,101 @@ func (s *State) ChannelRemove(channel *Channel) error {
 	return nil
 }
 
+// ThreadListSync syncs guild threads with provided ones.
+func (s *State) ThreadListSync(tls *ThreadListSync) error {
+	guild, err := s.Guild(tls.GuildID)
+	if err != nil {
+		return err
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	// This algorithm filters out archived or
+	// threads which are children of channels in channelIDs
+	// and then it adds all synced threads to guild threads and cache
+	index := 0
+	for _, t := range guild.Threads {
+		if !t.ThreadMetadata.Archived && tls.ChannelIDs != nil {
+			for _, v := range tls.ChannelIDs {
+				if t.ParentID == v {
+					goto remove
+				}
+			}
+			guild.Threads[index] = t
+			index++
+			continue
+		}
+	remove:
+		delete(s.channelMap, t.ID)
+	}
+	guild.Threads = guild.Threads[:index]
+	for _, t := range tls.Threads {
+		s.channelMap[t.ID] = t
+		guild.Threads = append(guild.Threads, t)
+	}
+
+	for _, m := range tls.Members {
+		if c, ok := s.channelMap[m.ID]; ok {
+			c.Member = m
+		}
+	}
+
+	return nil
+}
+
+// ThreadMembersUpdate updates thread members list
+func (s *State) ThreadMembersUpdate(tmu *ThreadMembersUpdate) error {
+	thread, err := s.Channel(tmu.ID)
+	if err != nil {
+		return err
+	}
+	s.Lock()
+	for idx, member := range thread.Members {
+		for _, removedMember := range tmu.RemovedMembers {
+			if member.ID == removedMember {
+				thread.Members = append(thread.Members[:idx], thread.Members[idx+1:]...)
+				break
+			}
+		}
+	}
+	s.Unlock()
+
+	for _, addedMember := range tmu.AddedMembers {
+		s.Lock()
+		thread.Members = append(thread.Members, addedMember.ThreadMember)
+		s.Unlock()
+		if addedMember.Member != nil {
+			err = s.MemberAdd(addedMember.Member)
+			if err != nil {
+				return err
+			}
+		}
+		if addedMember.Presence != nil {
+			err = s.PresenceAdd(tmu.GuildID, addedMember.Presence)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	s.Lock()
+	thread.MemberCount = tmu.MemberCount
+	s.Unlock()
+
+	return nil
+}
```

Comment:
This could easily add a member to the thread twice—if two updates happen concurrently, both could go through the deletion phase (the second one to execute wouldn't actually do anything), then both could go through the append phase. We really need one large lock on this entire method. I recognize that `MemberAdd` and `PresenceAdd` already take out locks, making this a bit more difficult—it might be worth adding unexported state methods that do the bulk of the work, and keep the exported methods as thin shims that take out the state lock. (Or, to be frank, it might be better to drop locking from the state altogether and have the caller handle the lock—currently, it's impossible to call multiple state functions in one locked sequence—but we might be too far down that road at this point.)